lib/types.nix

```diff
@@ -181,81 +247,165 @@ rec {
+      in
+      mkOptionType {
+        name = "one of ${concatStringsSep ", " (show values)}";
+        typerep = "(enum${concatMapStrings (x: "(${escape ["(" ")"] (builtins.toString x)})") values})";
+        check = x: elem x values && (builtins.isString x || builtins.isInt x);
+        merge = mergeOneOption;
+        nestedTypes = [];
+    };
+
+    either = t1: t2: mkOptionType {
+      name = "${t1.name} or ${t2.name}";
+      typerep = "(either${t1.typerep}${t2.typerep})";
+      check = x: t1.check x || t2.check x;
+      merge = mergeOneOption;
+      nestedTypes = t1.nestedTypes ++ t2.nestedTypes;
+      defaultValues = t1.defaultValues ++ t2.defaultValues;
     };
 
+
+    #
+    # Complex types
+    #
+
+    functionTo = resultType:
+      mkOptionType {
+        name = "function to ${resultType.name}";
+        typerep = "(function${resultType.typerep})";
+        check = builtins.isFunction;
+        merge = mergeOneOption;
+        nestedTypes = resultType.nestedTypes;
+        defaultValues = map (x: {...}:x) resultType.nestedTypes; # INFO: It seems as nix can't compare functions, yet.
+      };
+
     submodule = opts:
       let
         opts' = toList opts;
         inherit (import ./modules.nix) evalModules;
+        filterVisible = filter (opt: (if opt ? visible then opt.visible else true) && (if opt ? internal then !opt.internal else true));
       in
       mkOptionType rec {
         name = "submodule";
+        typerep = "(submodule)";
         check = x: isAttrs x || isFunction x;
-        merge = loc: defs:
+        merge = _module: loc: defs:
           let
+            internalModule = [ { inherit _module; } ];
             coerce = def: if isFunction def then def else { config = def; };
-            modules = opts' ++ map (def: { _file = def.file; imports = [(coerce def.value)]; }) defs;
+            modules = opts' ++ internalModule ++ map (def: { _file = def.file; imports = [(coerce def.value)]; }) defs;
           in (evalModules {
             inherit modules;
             args.name = last loc;
             prefix = loc;
           }).config;
-        getSubOptions = prefix: (evalModules
-          { modules = opts'; inherit prefix;
+        getSubOptions = getSubOptionsPrefixed [];
+        getSubOptionsPrefixed = prefix: _module:
+          let
+            internalModule = [ { inherit _module; } ];
+          in (evalModules {
+            modules = opts' ++ internalModule;
+            inherit prefix;
             # FIXME: hack to get shit to evaluate.
-            args = { name = ""; }; }).options;
+            args = { name = ""; };
+          }).options;
         getSubModules = opts';
         substSubModules = m: submodule m;
+        nestedTypes = concatMap (opt: opt.type.nestedTypes) (collect (lib.isType "option") (getSubOptions {}));
+        defaultValues = [{}];
       };
 
-    enum = values:
+    nixpkgsConfig = pkgs:
```

Comment:
I do not see any reason to move this type into this file.  As the name implies, this is the type of this unique option.